lib/factory.rb

```diff
@@ -14,3 +16,140 @@
 # - to_a
 # - values_at
 # - ==, eql?
+# ----------------------------------------------------------------------------
+class Factory
+  def self.new(*fields, &block)
+    raise ArgumentError if fields.empty?
+
+    fields.each do |field|
+      next if field.instance_of?(Symbol)
+
+      if field.instance_of?(String) && \
+         !field.empty? && \
+         field.match(/\A[A-Z]/)
+
+        next
+      end
+
+      raise ArgumentError, "identifier #{member} needs to be a constant"
+    end
+
+    identifier, *rest = fields
+
+    if identifier.instance_of? String
+      return const_set(identifier, create_class(*rest, &block))
+    end
+
+    create_class(*fields, &block)
+  end
+
+  def self.create_class(*fields, &block)
+    Class.new do
+      attr_accessor(*fields)
+
+      define_method :initialize do |*args|
+        raise ArgumentError, 'factory size differs' if fields.size < args.size
+
+        fields.each_with_index do |field, i|
```

Comment:
If you run `fasterer`, it will show that `each_with index` is slow. Let's do it with `zip` (zip fields with args and set variables)